src/btree/bt_read.c

```diff
@@ -257,7 +257,22 @@ __wt_page_in_func(WT_SESSION_IMPL *session, WT_REF *ref, uint32_t flags
             if (!LF_ISSET(WT_READ_IGNORE_CACHE_SIZE))
                 WT_RET(__wt_cache_eviction_check(
                   session, true, !F_ISSET(session->txn, WT_TXN_HAS_ID), NULL));
-            WT_RET(__page_read(session, ref, flags));
+
+            ret = __page_read(session, ref, flags);
+
+            /*
+             * Only check operation timeout if we fail to read the page.
+             *
+             * Don't check timeout for internal sessions as some operations like recovery and
+             * rollback to stable may take a long time and the default timeout may not be long
+             * enough. Also these operations are running with exclusive access and they should be
+             * able to continue and finish.
+             */
+            if (ret != 0) {
```

Comment:
The code can be better if we know the error code when reading a page fails because of cache full.

Reply:
Tested again without this part of the change and the rest of the change without it seems enough to make this problem not reproducing on the current develop. But I still think this code snippet is still necessary if we are timeout in the `__wt_cache_eviction_check ` ahead.